Add shim for Databricks 10.4 [databricks]

jlowe opened the pull request:
Closes #4914.

Adds a shim for the Databricks 10.4 runtime.  The dist pom has _not_ been updated to reference this version, as we need to setup build pipelines for these new Databricks-based jars.  Once those pipelines are setup, we can followup with a PR to update the dist pom and build scripts.

Some code from 301until330-all has been refactored into a 301until330-nondb directory to allow some reuse of files in the original 301until3300-all directory with the new 321db shim.

Once hiccup with this shim is that it appears `First` and possibly other aggregation expressions have an updated intermediate data format which lead to some aggregation test failures.  See #4963.  To be on the safe side until this is further investigated, aggregations are not replaced on Databricks 10.4 unless we can replace both sides of the aggregation.

jlowe commented:
build

jlowe commented:
build

revans2 reviewed (commented):
Generally this looks good to me, but I am not ready to approve it yet. There is just so much code I don't know what changes were copy/paste and what are specific to the new version of databricks. If you want to sit down and have us go over it together we can, or I can apply the patch and manually find the files with the same names and diff them.  Either way works fine with me.

jlowe commented on jenkins/databricks/build.sh:
```diff
@@ -68,80 +68,122 @@ CUDF_JAR=${M2DIR}/ai/rapids/cudf/${CUDF_VERSION}/cudf-${CUDF_VERSION}-${CUDA_VER
+        DB_VERSION=-0003
+        FASTERXML_JACKSON_VERSION=2.12.3
+        HADOOP_VERSION=3.2
+        HADOOP_FULL_VERSION=3.2.
```
Nice catch!  This was leftover from an attempt to reconcile jar dependencies where I thought we needed the full major.minor.patch version for the Hadoop jars, but then I ended up just pulling in the normal Hadoop client jar rather than worrying about using the exact version from the Databricks deploy.  If code compiled against the official Hadoop jars doesn't run on Databricks then that would be a real problem for that platform.

jlowe commented:
build

jlowe commented:
> I don't know what changes were copy/paste and what are specific to the new version of databricks.

I can give a quick overview of the majority of the changes.  They can be summed up in the following high-level changes:
- Moving some files in until330-all to until330-nondb, triggering copies in the other two existing Databricks shims
- New files in 321db (this new shim).  Most of these are copies of what the 321 shim was using, but there are exceptions.
- New code in AggregationTagging and aggregate.scala to support avoiding replacing only one half of an aggregate when using the 321db shim.

As expected, Databricks 10.4 is not just Apache Spark 3.2.1 plus custom changes, as it also has changes from Apache Spark 3.3.0 mixed in.  As such, there were files in '*until330-all' directories that no longer applied to all once this shim appeared.  To reconcile that, I moved the files incompatible with the new Databricks shim into until330-nondb directories which then triggered a copy of those files into the two existing Databricks shims that do not use the directory.  The 301db shim is going away very soon, so it's net one extra copy for the 312db shim.  Everything that was added to 301db and 31xdb should be the same as the new files added to corresponding until330-nondb directory.

Similarly, there was a 30Xuntil33X shim class that, once this shim was added, no longer applied to all shims.  I split out the incompatible methods into a new nondb shim and updated all the original users to use that as well.  The existing Databricks shims got copies of these separated methods since they don't use nondb dirs/code.

pxLi commented:
@NvTimLiu can you help take a look at the CICD requirement for db 10.4 shims? thanks!

tgravescs commented on integration_tests/src/main/python/array_test.py:
```diff
@@ -78,7 +78,7 @@ def test_array_item_with_strict_index(strict_index_enabled, index):
 @pytest.mark.parametrize('index', [-2, 100, array_neg_index_gen, array_out_index_gen], ids=idfn)
 def test_array_item_ansi_fail_invalid_index(index):
-    message = "java.lang.ArrayIndexOutOfBoundsException"
+    message = "SparkArrayIndexOutOfBoundsException" if is_databricks104_or_later() else "java.lang.ArrayIndexOutOfBoundsException"
```
these tests that check if is_databricks104_or_later are failing for DB 7.3 and 9.1, seems like its not using else clause

jlowe replied:
Yeah this is the string vs. numeric comparison issue on the version number.

revans2 commented on sql-plugin/src/main/321db/scala/com/nvidia/spark/rapids/shims/GpuHashPartitioning.scala:
```diff
@@ -0,0 +1,44 @@
+          expressions.length == h.expressions.length && expressions.zip(h.expressions).forall {
+            case (l, r) => l.semanticEquals(r)
+          }
+        case c @ ClusteredDistribution(requiredClustering, _) =>
```
nit: `c` is not used here. It would make the diff with the other versions of this file smaller.

revans2 commented on sql-plugin/src/main/330+/scala/org/apache/spark/sql/catalyst/json/rapids/shims/Spark33XFileOptionsShims.scala:
```diff
@@ -16,12 +16,12 @@
 import org.apache.spark.sql.catalyst.json.JSONOptions
 
-trait Spark33XFileOptionsShims extends Spark321PlusShims {
+trait Spark33XFileOptionsShims extends Spark321PlusShims with Spark320PlusNonDBShims {
```
nit: I personally think this would be better as an object in the v2 shims instead of a trait here. Before the names of the traits matched with what was expected, but now we have a new file for databricks 2.1 where the code matches this, but is named like the other ones.

revans2 commented on sql-plugin/src/main/321db/scala/com/nvidia/spark/rapids/shims/GpuRunningWindowExec.scala:
```diff
@@ -0,0 +1,39 @@
+/**
+ * GPU-based window-exec implementation, analogous to RunningWindowFunctionExec.
+ */
+class GpuRunningWindowExecMeta(runningWindowFunctionExec: RunningWindowFunctionExec,
```
nit: could we rename all of these files to GpuRunningWindowExecMeta.scala to match the naming convention for files?

revans2 commented on sql-plugin/src/main/321db/scala/com/nvidia/spark/rapids/shims/GpuRunningWindowExec.scala:
```diff
@@ -0,0 +1,39 @@
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids.shims
```
Would it be good for us to have a src director like 301+-db as only the copyright has changed from one file to the next for this?

revans2 commented on sql-plugin/src/main/321db/scala/com/nvidia/spark/rapids/shims/GpuWindowInPandasExec.scala:
```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
```
nit: for me if we are just copying a file I don't see a reason to update/modify the copyright.  If the file was written by someone other than NVIDIA we would not remove their copyright and add an NVIDIA one.

tgravescs replied:
I think our convention has been to update it. I'm not a lawyer and personally don't care as long as we say what the convention is and keep consistent.

jlowe replied:
The good news is that for this PR, I'm adding a 311+-db directory which removes quite a few redundant copies and nicely side-steps this issue.

jlowe commented on integration_tests/src/main/python/spark_session.py:
```diff
@@ -145,3 +145,7 @@ def is_spark_330_or_later():
+
+def is_databricks104_or_later():
+    spark = get_spark_i_know_what_i_am_doing()
+    return spark.conf.get("spark.databricks.clusterUsageTags.sparkVersion", "") >= "10.4"
```
This string comparison doesn't work as a numeric comparison.

revans2 replied:
You could zero lpad the spark conf.

jlowe replied:
Interesting, but it doesn't address the same issue happening in the minor version number.  I added code to split on `.` and convert the first two fields to integers which should resolve it.

jlowe commented on sql-plugin/src/main/321db/scala/org/apache/spark/sql/rapids/execution/python/shims/GpuFlatMapGroupsInPandasExec.scala:
```diff
@@ -0,0 +1,188 @@
+        StructField("out_struct", StructType.fromAttributes(localOutput)) :: Nil)
+
+    // Configs from DB 9.1 runtime
+    val maxBytes = conf.pandasZeroConfConversionMaxBytesPerSlice
```
The name of the fields in the conf changed in 10.4.  I updated the comment.

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala:
```diff
@@ -802,6 +802,26 @@ class GpuHashAggregateIterator(
+    currentMeta match {
+      case aggMeta: GpuBaseAggregateMeta[_] if aggMeta.agg.logicalLink.contains(logical) =>
+        List[GpuBaseAggregateMeta[_]](aggMeta) ++
+            getAggregateOfAllStages(aggMeta.childPlans.head, logical)
```
Normal indent is 2 and continuation indent is 4, so this looks properly indented to me and matches what the IDE wants using the recommended style settings.

tgravescs replied:
each level of indentation in scala style is 2 spaces.
https://docs.scala-lang.org/style/indentation.html

I think the only 4 space indentation is for parameters. (https://github.com/databricks/scala-style-guide#indent)

My IDE puts it at 2 spaces.

really in the example on the page I don't know you need any indentation and I've seen it similar in spark, but its by no means consistent, like: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L320

Its by no means a big deal.

jlowe replied:
Ah then we'll want to update https://github.com/NVIDIA/spark-rapids/blob/branch-22.04/docs/dev/idea-code-style-settings.xml#L28 accordingly, as I (and others) have been using 4-space continuation indent for quite some time now.

tgravescs replied:
https://github.com/NVIDIA/spark-rapids/issues/4986

jlowe commented on sql-plugin/src/main/321db/scala/org/apache/spark/rapids/shims/GpuShuffleExchangeExec.scala:
```diff
@@ -0,0 +1,74 @@
+  override val outputPartitioning: Partitioning = cpuOutputPartitioning
+
+  // 'mapOutputStatisticsFuture' is only needed when enable AQE.
+  override def mapOutputStatisticsFuture: Future[MapOutputStatistics] = {
```
I tried doing this but I get a compiler error because it fails to override. :shrug:

jlowe commented:
@tgravescs @revans2 I think I have addressed your concerns.

jlowe commented:
build

tgravescs commented on sql-plugin/src/main/321db/scala/com/nvidia/spark/rapids/shims/SparkShims.scala:
```diff
@@ -0,0 +1,187 @@
+
+  override def broadcastModeTransform(mode: BroadcastMode, rows: Array[InternalRow]): Any = {
+    val memoryManager = Option(TaskContext.get).map(_.taskMemoryManager()).getOrElse {
+      new TaskMemoryManager(SparkEnv.get.memoryManager, 0)
```
@jlowe ?

jlowe replied:
I added a comment earlier, so I went ahead and filed #4987 for further followup.

tgravescs replied:
great, sorry missed the comment before

tgravescs commented:
want to change out of draft?

jlowe commented:
> want to change out of draft?

Still waiting for the manual run of the integration tests on Datarbricks 10.4 on the latest changes to complete.  I expect it to pass as it did before, but I don't want this to be merged until we have confirmation those tests are passing.

jlowe commented:
Databricks 10.4 test run passed, so just waiting for the clean CI run at this point.
